### PR TITLE
`Tabs` component - Main feature branch (HDS-576)

### PR DESCRIPTION
**Pull request by KristinLBradley**
### :pushpin: Summary

If merged, this PR will add the Tabs component to the design system.

https://hds-components-git-hds-576-tabs-component-hashicorp.vercel.app/components/tabs

### :link: External links
https://hashicorp.atlassian.net/browse/HDS-576

***

### 👀 How to review

👉 Review commit-by-commit
👉 Review by files changed

Reviewer's checklist:

- [ ] +1 Percy if applicable
- [ ] Confirm that PR has a changelog update via [Changesets](https://github.com/changesets/changesets) if needed

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

### CODE REVIEW

**Comment by didoo** [comment]
@KristinLBradley I hope you don't mind if I rename the PR title, to differentiate from the other (otherwise I always get confused on which one the "main feature branch" one and which one is the WIP).

**Review by heatherlarsen — CHANGES_REQUESTED** [review]
Visually everything looks great 🎉 

But there are a few functionality things I noticed...

1. I noticed a difference between hitting `return` and hitting `space` when selecting a tab via the keyboard...When I keydown on `space` the focus ring disappears and then reappears on keyup. This doesn't happen when interacting with the `return` key. Is there a way to make these consistent so they both with like `return`?
2. I tried testing overflow by toggling the device to "iPhone SE" in Chrome's inspector but it doesn't look like the fallback scroll functionality is working correctly...it just overflows (_see screenshot_). I'd except this to have some sort of overflow functionality still.
Screen Shot 2022-10-06 at 11 43 11 AM

**Comment by didoo** [comment]
> 1. I noticed a difference between hitting `return` and hitting `space` when selecting a tab via the keyboard...When I keydown on `space` the focus ring disappears and then reappears on keyup. This doesn't happen when interacting with the `return` key. Is there a way to make these consistent so they both work like `return`?

@heatherlarsen that's because of this declaration:
screenshot_1853
that is shared across all the components (for the focus ring): when a `<button>` element (the only one that has an active state, links don't have it) is focused but also active (eg. you press the spacebar when it's focused), the focus ring disappears. That's how it works for buttons, for example: try here https://design-system-components-hashicorp.vercel.app/components/button and how it was required by designers (and in fact I added a note "by design")

> 2. I tried testing overflow by toggling the device to "iPhone SE" in Chrome's inspector but it doesn't look like the fallback scroll functionality is working correctly...it just overflows (_see screenshot_). I'd except this to have some sort of overflow functionality still.

@KristinLBradley I _think_ `.hds-tabs__tablist` needs an `overflow: auto` applied to it. Can you check?

**Comment by KristinLBradley** [comment]
> Visually everything looks great 🎉
> 
> But there are a few functionality things I noticed...
> 
> 1. I noticed a difference between hitting `return` and hitting `space` when selecting a tab via the keyboard...When I keydown on `space` the focus ring disappears and then reappears on keyup. This doesn't happen when interacting with the `return` key. Is there a way to make these consistent so they both work like `return`?
> 2. I tried testing overflow by toggling the device to "iPhone SE" in Chrome's inspector but it doesn't look like the fallback scroll functionality is working correctly...it just overflows (_see screenshot_). I'd except this to have some sort of overflow functionality still.
> 
> Screen Shot 2022-10-06 at 11 43 11 AM

@heatherlarsen  Thanks, I will test the keyboard functionality and see what I can do. 

I didn't merge in anything related to overflow yet to the main feature branch so that's why you see the content breaking out of the container like that. (I had tried some overflow styles previously in a sub-branch but undid them because Cristiano said they hadn't been reviewed before I merged.) I'm working on new simplified basic overflow styles in my new work in progress branch. I'll add you to the new branch once it's more ready.

**Comment by KristinLBradley** [comment]
> > 1. I noticed a difference between hitting `return` and hitting `space` when selecting a tab via the keyboard...When I keydown on `space` the focus ring disappears and then reappears on keyup. This doesn't happen when interacting with the `return` key. Is there a way to make these consistent so they both work like `return`?
> 
> @heatherlarsen that's because of this declaration: screenshot_1853 that is shared across all the components (for the focus ring): when a `<button>` element (the only one that has an active state, links don't have it) is focused but also active (eg. you press the spacebar when it's focused), the focus ring disappears. That's how it works for buttons, for example: try here https://design-system-components-hashicorp.vercel.app/components/button and how it was required by designers (and in fact I added a note "by design")
> 
> > 2. I tried testing overflow by toggling the device to "iPhone SE" in Chrome's inspector but it doesn't look like the fallback scroll functionality is working correctly...it just overflows (_see screenshot_). I'd except this to have some sort of overflow functionality still.
> 
> @KristinLBradley I _think_ `.hds-tabs__tablist` needs an `overflow: auto` applied to it. Can you check?

@didoo Yes, I had undone what I tried before so hadn't merged any styles related to overflow behavior yet. I'm adding styles for those in my new WIP sub-branch for the "polish".

**Review by didoo — COMMENTED** [review]
@KristinLBradley some small changes and then we're probably good to go

(there's one last thing I want to investigate, but is not a blocker, but I'll do it tomorrow)

File: packages/components/addon/components/hds/tabs/panel.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/panel.hbs]: Thanks @didoo. 

@MelSumner please let me know if it's better to move these, thanks.

File: packages/components/addon/components/hds/tabs/panel.js
Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/panel.js:13]: Ok this was a comment from one of the reviews I added. I'll commit your deletion, thanks.

File: packages/components/addon/components/hds/tabs/tab.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/tab.hbs:2]: Yes, it works, thanks

File: packages/components/addon/components/hds/tabs/tab.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/tab.hbs:2]: Thanks. 

@MelSumner is it generally a good idea to move these after `...attributes`?

**Review by heatherlarsen — APPROVED** [review]
This looks great! Wonderful work making it come to life 👏 👏 👏

**Review by didoo — APPROVED** [review]
Apart from waiting for the feedback from @MelSumner about moving (or not) some attributes after the `...attributes` declaration, the PR is good to go for me.

I 100% agree with Heather's comment and praise you for the wonderful work 🙌

File: packages/components/addon/components/hds/tabs/panel.js
Comment (Dhaulagiri) [reply packages/components/addon/components/hds/tabs/panel.js]: Unsure if the "can" here is related to the functionality or this feature actually existing, but FWIW we are on a version of ember that does include this now and there is also another [polyfill](https://www.npmjs.com/package/ember-unique-id-helper-polyfill) to backport this for all our consumers who are not on the later versions of ember.

Comment (didoo) [reply packages/components/addon/components/hds/tabs/panel.js]: @Dhaulagiri with _"we can't use here"_ I meant in JS code (unless I am missing something: looking at the RFC to me seems it can be used only in templates, not in backing classes).
In the specific case of the tabs there are two things that makes me think we can't use it as is:
- we manipulate the `unique-id` (we add a prefix to it)
- we need it on the JS side of things: how would you access the value of `{{unique-id}}` in JS? I can't find references about it.

Comment (Dhaulagiri) [reply packages/components/addon/components/hds/tabs/panel.js]: 🙇  thanks for the explanation. seems prudent to ignore me here 😅

File: packages/components/addon/components/hds/tabs/panel.hbs
Comment (MelSumner) [reply packages/components/addon/components/hds/tabs/panel.hbs]: @KristinLBradley Here's the docs: https://guides.emberjs.com/release/in-depth-topics/patterns-for-components/#toc_attribute-ordering 

I agree that `...attributes` in this case would probably need to be right below the `class` attribute. 👍

File: packages/components/addon/components/hds/tabs/tab.hbs
Comment (MelSumner) [reply packages/components/addon/components/hds/tabs/tab.hbs:2]: @KristinLBradley yes, I'd move it to before `role`

**Review by MelSumner — CHANGES_REQUESTED** [review]
The `...attributes` needs to be moved in a couple of cases but other than that it looks good!

File: packages/components/addon/components/hds/tabs/panel.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/panel.hbs]: Thanks!

**Review by MelSumner — APPROVED** [review]

File: .changeset/tall-rice-attend.md
Comment (alex-ju) [inline .changeset/tall-rice-attend.md:5]: suggestion: we probably don't need quotes on the changelog entries

**Review by alex-ju — APPROVED** [review]
Looks good overall! I am aware this code has already been reviewed in the PRs that were raised against this main feature branch, so did a thin review and left a few thoughts here not to forget about them.

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/index.hbs]: Thanks. I renamed these 2:
```
onClick=this.onClick
onKeyUp=this.onKeyUp
```

Comment (alex-ju) [reply packages/components/addon/components/hds/tabs/index.hbs]: Ah, lovely! Thank you!

**Pull request merged**
